geode-core/src/test/java/org/apache/geode/internal/cache/map/RegionMapPutTest.java

```diff
@@ -89,13 +106,307 @@ private RegionEntry doPut() {
     return instance.put();
   }
 
+  @Test
+  public void doesNotSetEventOldValueToExistingRegionEntryValue_ifNotRequired() {
+    givenExistingRegionEntry();
+    givenAnOperationThatDoesNotGuaranteeOldValue();
+    givenPutDoesNotNeedToDoCacheWrite();
+    this.requireOldValue = false;
+
+    Object oldValue = new Object();
+    when(existingRegionEntry.getValue()).thenReturn(oldValue);
+
+    doPut();
+
+    verify(event, never()).setOldValue(any());
+    verify(event, never()).setOldValue(any(), anyBoolean());
+  }
+
+  @Test
+  public void setsEventOldValueToExistingRegionEntryValue_ifOldValueIsGatewaySenderEvent() {
+    givenExistingRegionEntry();
+
+    GatewaySenderEventImpl oldValue = new GatewaySenderEventImpl();
+    when(existingRegionEntry.getValue()).thenReturn(oldValue);
+
+    doPut();
+
+    verify(event, times(1)).setOldValue(same(oldValue), eq(true));
+    verify(event, never()).setOldValue(not(same(oldValue)), eq(true));
+  }
+
+  @Test
+  public void setsEventOldValueToExistingRegionEntryValue_ifOperationGuaranteesOldValue() {
+    givenExistingRegionEntry();
+    givenAnOperationThatGuaranteesOldValue();
+
+    Object oldValue = new Object();
+    when(existingRegionEntry.getValueOffHeapOrDiskWithoutFaultIn(same(internalRegion)))
+        .thenReturn(oldValue);
+
+    doPut();
+
+    verify(event, times(1)).setOldValue(same(oldValue), eq(true));
+    verify(event, never()).setOldValue(not(same(oldValue)), eq(true));
+  }
+
+  @Test
+  public void eventPutExistingEntryGivenOldValue_ifRetrieveOldValueForDelta()
+      throws RegionClearedException {
+    givenExistingRegionEntry();
+    Object oldValue = new Object();
+    when(existingRegionEntry.getValue(any())).thenReturn(oldValue);
+    when(event.getDeltaBytes()).thenReturn(new byte[1]);
+    doAnswer(invocation -> {
+      Runnable runnable = invocation.getArgument(0);
+      runnable.run();
+      return null;
+    }).when(focusedRegionMap).runWhileEvictionDisabled(any());
+
```

Comment:
Consider extracting this lambda as a method:

```
  private static Object runFirstArgument(InvocationOnMock invocation) {
    Runnable runnable = invocation.getArgument(0);
    runnable.run();
    return null;
  }
```

Then you could pass it via a method reference in this test and the next one:

        doAnswer(RegionMapPutTest::runFirstArgument)
        .when(focusedRegionMap).runWhileEvictionDisabled(any());

If you can think of a better name than `runFirstArgument`, please do ;-)

Reply:
I actually prefer the lambda in this case because it is such a localized, one off, action we want to do. We are just trying to mock "runWhileEvictionDisabled" to call the runnable it is passed. So if we extracted the lambda into its own method it would be hard to understand that method out of the context of the caller.
But I didn't like the noise and the duplication of setup code these two tests have so I extracted this block of code into a method named: givenThatRunWhileEvictionDisabledCallsItsRunnable

Reply:
Oh, that's way better than what I said.